src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts

```diff
@@ -627,284 +582,148 @@ export class TreeGrid<DATA>
-            }).catch((reason: any) => {
-                this.initData([]);
-                this.handleError(reason);
-            }).then(() => {
-                this.updateExpanded();
-            }).then(() => this.notifyLoaded());
+        return this.doExpandNode(this.root.getCurrentRoot()).then(() => {
+            if (this.hasHighlightedNode()) {
+                this.highlightCurrentNode();
+            }
+        }).catch((reason: any) => {
+            this.initData([]);
+            this.handleError(reason);
+        }).then(() => {
+            this.setActive(true);
+            this.notifyLoaded();
+        });
     }
 
-    refreshNode(node?: TreeNode<DATA>): void {
-        let root = this.root.getCurrentRoot();
-        this.setActive(false);
-
-        node = node || root;
-        node.regenerateIds();
-        root.setExpanded(true);
-        this.initData(root.treeToList());
-
-        this.invalidate();
-
-        this.setActive(true);
-
-        this.notifyLoaded();
+    updateNodesByData(dataItems: DATA[]) {
+        dataItems.forEach((updatedData: DATA) => {
+            this.updateNodeByData(updatedData);
+        });
     }
 
-    // Soft reset, that saves node status
-    refresh(): void {
-        let root = this.root.getCurrentRoot();
-
-        this.setActive(false);
-
-        this.grid.invalidate();
-
-        root.setExpanded(true);
-        this.initData(root.treeToList());
-        this.invalidate();
+    updateNodeByData(updatedData: DATA, oldDataId?: string) {
+        const dataId: string = !!oldDataId ? oldDataId : updatedData.getId();
+        const treeNodes: TreeNode<DATA>[] = this.root.getNodesByDataId(dataId);
 
-        this.setActive(true);
-
-        this.notifyLoaded();
+        treeNodes.forEach((updatedNode: TreeNode<DATA>) => {
+            updatedNode.setData(updatedData);
+            this.invalidateNodes([updatedNode]);
+            this.notifyDataChanged(new DataChangedEvent<DATA>([updatedNode], DataChangedType.UPDATED));
+        });
     }
 
-    updateNode(data: DATA, oldDataId?: string): Q.Promise<void> {
-
-        let dataId = oldDataId || this.getDataId(data);
-        let nodeToUpdate = this.root.getCurrentRoot().findNode(dataId);
+    deleteNodeByDataId(dataId: string) {
+        const nodes: TreeNode<DATA>[] = this.root.getNodesByDataId(dataId);
 
-        if (!nodeToUpdate) {
-            throw new Error('TreeNode to update not found: ' + dataId);
-        }
-
-        return this.fetchAndUpdateNodes([nodeToUpdate], oldDataId ? this.getDataId(data) : undefined);
+        nodes.forEach((node: TreeNode<DATA>) => {
+            this.deleteNode(node);
+        });
     }
 
-    updateNodeByDataId(dataId: string): Q.Promise<void> {
-        const nodesToUpdate: TreeNode<DATA>[] = this.root.getNodesByDataId(dataId);
-
-        if (nodesToUpdate.length > 0) {
-            return this.fetchAndUpdateNodes(nodesToUpdate);
+    protected deleteNode(node: TreeNode<DATA>): void {
```

Comment:
Codacy found an issue: [Method 'deleteNode' has a complexity of 7. Maximum allowed is 4.](https://app.codacy.com/manual/enonic/lib-admin-ui/pullRequest?prid=6422415)